src/solvers/refinement/string_refinement.cpp

```diff
@@ -882,8 +882,13 @@ exprt string_refinementt::substitute_array_with_expr(
     exprt else_expr=substitute_array_with_expr(with_expr.old(), index);
     const typet &type=then_expr.type();
     CHECK_RETURN(else_expr.type()==type);
+#if 0
+    // this does not hold for some reason
```

Comment:
actually the statement was wrong, it should have been `index.type()==with_expr.where().type`. I fixed this and now it works.